grpc/src/main/java/io/stargate/grpc/service/MessageHandler.java

```diff
@@ -0,0 +1,451 @@
+  protected final Persistence persistence;
+  private final StreamObserver<Response> responseObserver;
+
+  protected MessageHandler(
+      MessageT message,
+      Connection connection,
+      Cache<PrepareInfo, CompletionStage<Prepared>> preparedCache,
+      Persistence persistence,
+      StreamObserver<Response> responseObserver) {
+    this.message = message;
+    this.connection = connection;
+    this.preparedCache = preparedCache;
+    this.persistence = persistence;
+    this.responseObserver = responseObserver;
+  }
+
+  void handle() {
+    CompletionStage<Result> resultFuture =
+        CompletableFuture.<Void>completedFuture(null)
+            .thenApply(this::invokeValidate)
+            .thenCompose(__ -> prepare(false))
+            .thenCompose(this::executePrepared);
+    handleUnprepared(resultFuture)
+        .thenApply(this::buildResponse)
+        .thenCompose(this::executeTracingQueryIfNeeded)
+        .whenComplete(
+            (response, error) -> {
+              if (error != null) {
+                handleException(error);
+              } else {
+                setSuccess(response);
+              }
+            });
+  }
+
+  /** Performs any necessary validation on the message before execution starts. */
+  protected abstract void validate() throws Exception;
+
+  /**
+   * Prepares any CQL query required for the execution of the request, and returns an executable
+   * object.
+   *
+   * @param shouldInvalidate whether to invalidate the corresponding entries in the prepared
+   *     statement cache.
+   */
+  protected abstract CompletionStage<PreparedT> prepare(boolean shouldInvalidate);
+
+  /** Executes the prepared object to get the CQL results. */
+  protected abstract CompletionStage<Result> executePrepared(PreparedT prepared);
+
+  /** Builds the gRPC response from the CQL result. */
+  protected abstract ResponseAndTraceId buildResponse(Result result);
+
+  /** Computes the consistency level to use for tracing queries. */
+  protected abstract ConsistencyLevel getTracingConsistency();
+
+  private Void invokeValidate(Void ignored) {
+    try {
+      validate();
+      return null;
+    } catch (RuntimeException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new CompletionException(e);
+    }
+  }
+
+  protected BoundStatement bindValues(
+      PayloadHandler handler, Prepared prepared, QueryOuterClass.Payload values) throws Exception {
+    return values.hasData()
+        ? handler.bindValues(prepared, values.getData(), persistence.unsetValue())
+        : new BoundStatement(prepared.statementId, Collections.emptyList(), null);
+  }
+
+  protected CompletionStage<Prepared> prepare(PrepareInfo prepareInfo, boolean shouldInvalidate) {
+    // In the event a query is being retried due to a PreparedQueryNotFoundException invalidate the
+    // local cache to refresh with the remote cache
+    if (shouldInvalidate) {
+      preparedCache.invalidate(prepareInfo);
```

Comment:
Any worry about invalidating entries we just replaced?

Reply:
I see, one problematic scenario would be:
```
[thread1] prepare, cache hit
[thread1] execute, receive UNPREPARED
[thread2] prepare, cache hit
[thread2] execute, receive UNPREPARED
[thread1] invalidate, put new future in cache, reprepare
[thread2] invalidate, put new future in cache, reprepare ==> wasteful, new entry was already there
```

We could keep a reference to the entry we got on the first hit, and only invalidate if it's still in the cache by then.

But maybe that's a bit overkill, it complicates the code, and repreparing twice is just slightly wasteful, it does not introduce any bug.